# MC316 Final Exam – Part B: PYTHON

* You will have 1 hour and 40 minutes to finish the test.
* This part includes 10 questions. Each question is worth 15 points.
* You get extra 30 points for full answers.
* Please read the instructions carefully. 

## I. SET UP


### Import pandas library for use in scripting. 

In [157]:
import pandas as pd

### Import the datasets and name them.
For this test, you will analyze revenue and expenditure data released by Kansas Department of Administration for the fiscal year 2018. While the original data can be directly downloaded [here](http://kanview.ks.gov/DataDownload.aspx), for your convenience, the instructor had them uploaded to a server at the following urls. 

#### For Kansas State University:
1. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv
2. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv

#### For University of Kansas:
3. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv
4. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv

You will need to import all 4 datasets and name them in an easy way to call out later.

*Please refer to the [Glossary](http://kanview.ks.gov/Glossary.aspx#f) for data description.

In [158]:
KSU1 = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv")

In [159]:
KSU2 = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv")

In [160]:
KU1 = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv")

In [161]:
KU2 = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv")

## II. REQUIRED QUESTIONS
* Make sure you type the answers in the cells right below the codes.
* If you need to add extra cells, click the + button in the toolbar up the top. 

### Q1: What is the total revenue recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total revenue recorded for University of Kansas in the same year?

In [162]:
KU2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1077 entries, 0 to 1076
Data columns (total 22 columns):
Fiscal_Year                        1077 non-null int64
Fiscal_Quarter                     1077 non-null int64
Business_Unit                      1077 non-null int64
Business_Unit_Descr                1077 non-null object
GAAP_Fund_Type                     1077 non-null int64
GAAP_Fund_Type_Descr               1077 non-null object
Fund_Code                          1077 non-null int64
Fund_Code_Descr                    1077 non-null object
Budget_Ref                         1077 non-null int64
Budget_Ref_Descr                   1077 non-null object
Program_Code                       1077 non-null int64
Program_Code_Descr                 1077 non-null object
State_Function                     1077 non-null int64
State_Function_Descr               1077 non-null object
Primary_Account_Code               1077 non-null int64
Primary_Account_Code_Descr         1077 non-null object
Interm

In [163]:
KSU2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1910 entries, 0 to 1909
Data columns (total 22 columns):
Fiscal_Year                        1910 non-null int64
Fiscal_Quarter                     1910 non-null int64
Business_Unit                      1910 non-null int64
Business_Unit_Descr                1910 non-null object
GAAP_Fund_Type                     1910 non-null int64
GAAP_Fund_Type_Descr               1910 non-null object
Fund_Code                          1910 non-null int64
Fund_Code_Descr                    1910 non-null object
Budget_Ref                         1910 non-null int64
Budget_Ref_Descr                   1910 non-null object
Program_Code                       1910 non-null int64
Program_Code_Descr                 1910 non-null object
State_Function                     1910 non-null int64
State_Function_Descr               1910 non-null object
Primary_Account_Code               1910 non-null int64
Primary_Account_Code_Descr         1910 non-null object
Interm

In [164]:
KSU2.Total.sum()

854324643.84

In [165]:
KU2.Total.sum()

808767902.55

Answer: The total revenue cost for Kansas State University in the Fiscal Year of 2018 is 854324643.84 and compared to Kansas University it's 808767902.55 in that same year.

### Q2: What is the total expense recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total expense recorded for University of Kansas in the same year?

In [166]:
KSU1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24094 entries, 0 to 24093
Data columns (total 22 columns):
Fiscal_Year                        24094 non-null int64
Fiscal_Quarter                     24094 non-null int64
Business_Unit                      24094 non-null int64
Business_Unit_Descr                24094 non-null object
GAAP_Fund_Type                     24094 non-null int64
GAAP_Fund_Type_Descr               24094 non-null object
Fund_Code                          24094 non-null int64
Fund_Code_Descr                    24094 non-null object
Budget_Ref                         24094 non-null int64
Budget_Ref_Descr                   24094 non-null object
Program_Code                       24094 non-null int64
Program_Code_Descr                 24094 non-null object
State_Function                     24094 non-null int64
State_Function_Descr               24094 non-null object
Primary_Account_Code               24094 non-null int64
Primary_Account_Code_Descr         24094 non-

In [167]:
KU1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10553 entries, 0 to 10552
Data columns (total 22 columns):
Fiscal_Year                        10553 non-null int64
Fiscal_Quarter                     10553 non-null int64
Business_Unit                      10553 non-null int64
Business_Unit_Descr                10553 non-null object
GAAP_Fund_Type                     10553 non-null int64
GAAP_Fund_Type_Descr               10553 non-null object
Fund_Code                          10553 non-null int64
Fund_Code_Descr                    10553 non-null object
Budget_Ref                         10553 non-null int64
Budget_Ref_Descr                   10553 non-null object
Program_Code                       10553 non-null int64
Program_Code_Descr                 10553 non-null object
State_Function                     10553 non-null int64
State_Function_Descr               10553 non-null object
Primary_Account_Code               10553 non-null int64
Primary_Account_Code_Descr         10553 non-

In [168]:
KSU1.Total.sum()

1018584373.02

In [169]:
KU1.Total.sum()

990238875.6099999

Answer: The total expense for Kansas State University in the Fiscal Year of 2018 is 1018584373.02 and compared to Kansas University is 990238875.6099999 in that same year.

### Q3: Where did the two universities' primary revenues come from? 
*Refer to the description [here](http://kanview.ks.gov/Revenues/AcctCodeRevenues.aspx) to write your answer.

In [170]:
KSU2.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   637
REV FROM USE OF MONEY & PROP      354
REVENUE TRANSFERS                 341
NON REVENUE RECEIPTS              259
OTHER REVENUE                     160
GIFT DONATION FEDERAL GRANTS      152
TAXES                               7
Name: Primary_Account_Code_Descr, dtype: int64

In [171]:
KU2.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                   376
REV FROM USE OF MONEY & PROP      251
NON REVENUE RECEIPTS              157
REVENUE TRANSFERS                 153
OTHER REVENUE                      57
TAXES                              45
GIFT DONATION FEDERAL GRANTS       38
Name: Primary_Account_Code_Descr, dtype: int64

Answer: The two universities primary revenues come from Agency Earnings.

### Q4: What was the biggest source of revenues for Kansas State University? Compare to the one of University of Kansas.

Remove hashtag and run the following code to suppress scientific notation:

In [172]:
pd.set_option('display.float_format', lambda x: '%.3f' % x) 

In [173]:
KSU2.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  412565112.820
GIFT DONATION FEDERAL GRANTS     278778848.030
NON REVENUE RECEIPTS              71947564.050
OTHER REVENUE                      8432030.160
REV FROM USE OF MONEY & PROP      34139318.040
REVENUE TRANSFERS                 48458616.490
TAXES                                 3154.250
Name: Total, dtype: float64

In [174]:
KU2.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  408409337.320
GIFT DONATION FEDERAL GRANTS     193836668.320
NON REVENUE RECEIPTS              69753247.670
OTHER REVENUE                     68241930.500
REV FROM USE OF MONEY & PROP      40950779.540
REVENUE TRANSFERS                 26013855.400
TAXES                              1562083.800
Name: Total, dtype: float64

Answer: The biggest source of revenues for Kansas State University is Agency Earnings 412565112.820 and compared to Kansas University it is 408409337.320.

### Q5: What accounts contributed to Kansas State University's biggest source of revenues?
If you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [175]:
KSU2["Primary_Account_Code_Descr"] = KU2["Primary_Account_Code_Descr"].str.strip()

After cleaning the spaces, you can start filtering your datasets to find the answer for Q5.

In [176]:
KSU2.Primary_Account_Code_Descr.value_counts()

AGENCY EARNINGS                 376
REV FROM USE OF MONEY & PROP    251
NON REVENUE RECEIPTS            157
REVENUE TRANSFERS               153
OTHER REVENUE                    57
TAXES                            45
GIFT DONATION FEDERAL GRANTS     38
Name: Primary_Account_Code_Descr, dtype: int64

In [177]:
KSU2.Primary_Account_Code_Descr == "AGENCY EARNINGS"

0        True
1       False
2       False
3        True
4        True
5       False
6       False
7       False
8       False
9       False
10      False
11       True
12      False
13      False
14      False
15      False
16      False
17      False
18       True
19      False
20      False
21      False
22      False
23      False
24      False
25       True
26       True
27       True
28       True
29      False
        ...  
1880    False
1881    False
1882    False
1883    False
1884    False
1885    False
1886    False
1887    False
1888    False
1889    False
1890    False
1891    False
1892    False
1893    False
1894    False
1895    False
1896    False
1897    False
1898    False
1899    False
1900    False
1901    False
1902    False
1903    False
1904    False
1905    False
1906    False
1907    False
1908    False
1909    False
Name: Primary_Account_Code_Descr, Length: 1910, dtype: bool

In [178]:
KSU2[KSU2.Primary_Account_Code_Descr == "AGENCY EARNINGS"]

Fiscal_Year  Fiscal_Quarter  Business_Unit  \
0            2018               1            367   
3            2018               1            367   
4            2018               1            367   
11           2018               1            367   
18           2018               1            367   
25           2018               1            367   
26           2018               1            367   
27           2018               1            367   
28           2018               1            367   
37           2018               1            367   
38           2018               1            367   
39           2018               1            367   
40           2018               1            367   
41           2018               1            367   
42           2018               1            367   
43           2018               1            367   
44           2018               1            367   
54           2018               1            367   
55           2018               1            367   
56           2018               1            367   
57           2018               1            367   
58           2018               1            367   
59           2018               1            367   
65           2018               1            367   
66           2018               1            367   
67           2018               1            367   
68           2018               1            367   
69           2018               1            367   
76           2018               1            367   
77           2018               1            367   
...           ...             ...            ...   
973          2018               3            367   
974          2018               3            367   
975          2018               3            367   
976          2018               3            367   
983          2018               3            367   
985          2018               3            367   
986          2018               3            367   
987          2018               3            367   
989          2018               3            367   
991          2018               3            367   
992          2018               3            367   
1002         2018               3            367   
1003         2018               3            367   
1004         2018               3            367   
1008         2018               3            367   
1012         2018               3            367   
1013         2018               3            367   
1014         2018               3            367   
1015         2018               3            367   
1019         2018               3            367   
1023         2018               3            367   
1034         2018               3            367   
1035         2018               3            367   
1036         2018               3            367   
1057         2018               3            367   
1061         2018               3            367   
1071         2018               3            367   
1073         2018               3            367   
1075         2018               3            367   
1076         2018               3            367   

                 Business_Unit_Descr  GAAP_Fund_Type  \
0     Kansas State University                      1   
3     Kansas State University                     12   
4     Kansas State University                     12   
11    Kansas State University                     12   
18    Kansas State University                     12   
25    Kansas State University                     12   
26    Kansas State University                     12   
27    Kansas State University                     12   
28    Kansas State University                     12   
37    Kansas State University                     12   
38    Kansas State University                     12   
39    Kansas State University                     12   
40    Kansas State University                     12 

In [179]:
rev = KSU2[KSU2.Primary_Account_Code_Descr == "AGENCY EARNINGS"]

In [180]:
rev.head()

Fiscal_Year  Fiscal_Quarter  Business_Unit  \
0          2018               1            367   
3          2018               1            367   
4          2018               1            367   
11         2018               1            367   
18         2018               1            367   

               Business_Unit_Descr  GAAP_Fund_Type  \
0   Kansas State University                      1   
3   Kansas State University                     12   
4   Kansas State University                     12   
11  Kansas State University                     12   
18  Kansas State University                     12   

              GAAP_Fund_Type_Descr  Fund_Code                 Fund_Code_Descr  \
0   General Fund                         1000  STATE GENERAL FUND               
3   Component Unit                       2062  GENERAL FF                       
4   Component Unit                       2062  GENERAL FF                       
11  Component Unit                       2129  GENERAL FF                       
18  Component Unit                       2484  DEFERRED MNT SUPPORT FD          

    Budget_Ref                                   Budget_Ref_Descr  ...  \
0            3  OPERATING EXP-INCLD OFF HOS                   ...  ...   
3         2000  GENERAL FF                                    ...  ...   
4         2000  GENERAL FF                                    ...  ...   
11        5500  GENERAL FF                                    ...  ...   
18        2484  DEFERRED MNT SUPPORT FD                       ...  ...   

    State_Function                               State_Function_Descr  \
0                3  Education                                     ...   
3               26  Universities                                  ...   
4               26  Universities                                  ...   
11              26  Universities                                  ...   
18              26  Universities                                  ...   

    Primary_Account_Code Primary_Account_Code_Descr  \
0                   4300            AGENCY EARNINGS   
3                   4300            AGENCY EARNINGS   
4                   4600            AGENCY EARNINGS   
11                  7600            AGENCY EARNINGS   
18                  4300            AGENCY EARNINGS   

    Intermediate_Account_Code Intermediate_Account_Code_Descr  \
0                       43000  INTEREST AND INVESTMENTS         
3                       43000  INTEREST AND INVESTMENTS         
4                       46900  OTHER NON REVENUE RECEIPTS       
11                      76000  REVENUE TRANSFERS                
18                      43000  INTEREST AND INVESTMENTS         

    Detail_Account_Code       Detail_Account_Code_Descr     Total Unnamed: 21  
0                430230  INT AGENCY BANK ACCOUNTS         144.360         nan  
3                430150  AVERAGE DAILY BALANCE INTEREST 13821.250         nan  
4                469090  OTHER NONREVENUE RECEIPTS        105.870         nan  
11               766080  OP TRANSFER OUT INTEREST ALLOC -2016.110         nan  
18               430150  AVERAGE DAILY BALANCE INTEREST   277.140         nan  

[5 rows x 22 columns]

In [181]:
rev.groupby("Detail_Account_Code_Descr").Total.sum()

Detail_Account_Code_Descr
ADMISSIONS TO PLAYS & CONCERTS     919425.540
AGY FUNDS AUTHORIZED DEBT SVC     2945413.290
ALL OTHER OPERATING GRANTS        3902070.940
AUDITS AND INSPECTIONS              40077.020
AVERAGE DAILY BALANCE INTEREST      35664.740
CARE AND HOSPITALIZATION OTHER    1082519.020
CLERICAL SERVICES                  249948.770
CONVENIENCE FEE CREDIT CARDS        19232.000
CORPORATION SERVICES               153614.900
DEPARTMENTAL OR AGENCY SALES      2442689.820
EDUCATION AND LIBRARIES          19626740.470
FARM AND DAIRY PRODUCTS            433561.960
FED SUBGRANT TRANSFER IN           389567.540
FEDERAL GRANT OPERATING          27448441.970
INT AGENCY BANK ACCOUNTS              144.360
MANUFACTURED PRODUCTS                 315.000
MEALS AND PROCESSED FOODS         1928917.360
OIL GAS MINERAL SAND ROYALTIES      13621.330
OP TRANSFER IN INTEREST ALLOC      352038.900
OP TRANSFER OUT INTEREST ALLOC     -46628.010
OPERATING TRANSFERS IN            3199270.030
OPERATIN

Answer: See the table above

### Q6: Where did the money go? (Answer for both universities)
*Refer to the description [here] (http://kanview.ks.gov/Expenditures/AcctCodeExpenditures.aspx) to write your answer.

In [182]:
KSU1.Primary_Account_Code_Descr.value_counts()

CONTRACTUAL SERVICES              10561
SALARIES AND WAGES                 7780
COMMODITIES                        3740
CAPITAL OUTLAY                     1173
GRANTS CLAIMS SHARED REVENUE        534
NONEXPENSE ITEMS                    129
EXPENSE TRANSFERS                   106
DEBT SERVICE                         71
Name: Primary_Account_Code_Descr, dtype: int64

In [183]:
KU1.Primary_Account_Code_Descr.value_counts()

CONTRACTUAL SERVICES              4322
SALARIES AND WAGES                3623
COMMODITIES                       1605
CAPITAL OUTLAY                     737
NONEXPENSE ITEMS                    81
GRANTS CLAIMS SHARED REVENUE        75
EXPENSE TRANSFERS                   66
DEBT SERVICE                        44
Name: Primary_Account_Code_Descr, dtype: int64

Answer: See the table above. 

### Q7: What costs two universities the most? 

In [184]:
KSU1.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
CAPITAL OUTLAY                    64350550.130
COMMODITIES                       42670908.840
CONTRACTUAL SERVICES             128541598.280
DEBT SERVICE                      32442445.100
EXPENSE TRANSFERS                 37130576.230
GRANTS CLAIMS SHARED REVENUE     106551248.640
NONEXPENSE ITEMS                 130529992.790
SALARIES AND WAGES               476367053.010
Name: Total, dtype: float64

In [155]:
KU1.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
CAPITAL OUTLAY                    70676545.250
COMMODITIES                       24225857.120
CONTRACTUAL SERVICES             139630507.780
DEBT SERVICE                      25309742.030
EXPENSE TRANSFERS                 25815417.520
GRANTS CLAIMS SHARED REVENUE      71439705.360
NONEXPENSE ITEMS                 161145084.400
SALARIES AND WAGES               471996016.150
Name: Total, dtype: float64

Answer: Salaries and wages

### Q8: Calculate the percentage of this cost over each university's total expense.

In [187]:
476367053.010 / 471996016.150

1.0092607494776205

<b>Answer:</b> 

### Q9: Breakdown the cost you found in Q7 for both universities.
Again, if you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [3]:
#yourdataframe["yourcolumnname"] = yourdataframe["yourcolumnname"].str.strip()

<b>Answer:</b>

### Q10: Compare the breakdown costs from Q9 side by side.
When you merge the two subsets, make sure to add the attribute <b>how ="outer"</b> so that you can keep every item from both sets.
*For example: merge = pd.merge(df1, df2, how="outer", on="columnname")

<b>Answer:</b>

## III. BONUS QUESTIONS
* Each question's worth 15 points.


### B1: Which university made a profit in the fiscal year of 2018?

<b>Answer:</b>

### B2: Which software program do you like the most in this course? Why?

Answer: I like SQLite because it's basically straightforward. You're not going too in depth in the codes compared to the other softwares we've used this semester.

### END OF THE TEST